File: src/stack/min-stack.ts
// DIFFICULTY: MEDIUM
//
// Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
//
// Implement the MinStack class:
//
// MinStack() initializes the stack object.
// void push(int val) pushes the element val onto the stack.
// void pop() removes the element on the top of the stack.
// int top() gets the top element of the stack.
// int getMin() retrieves the minimum element in the stack.
//
// You must implement a solution with O(1) time complexity for each function.
//
// See {@link https://leetcode.com/problems/min-stack/}
export { MinStack };

// SOLUTION:
class MinStack {
  private readonly stack: number[];

  private readonly min: number[];

  constructor() {
    this.stack = [];

    // This stack will contain the same number of elements as the underlying stack, but each element pushed onto it is
    // the minimum value element at the time.
    this.min = [];
  }

  push(x: number): void {
    this.stack.push(x);

    if (this.min.length === 0) {
      this.min.push(x);
      return;
    }

    const smallest = Math.min(x, this.getMin());
    this.min.push(smallest);
  }

  pop(): void {
    this.stack.pop();
    this.min.pop();
  }

  top(): number {
    return this.stack[this.stack.length - 1];
  }

  getMin(): number {
    if (this.min.length === 0) {
      return -Infinity;
    }

    return this.min[this.min.length - 1];
  }
}

File: src/leetcode/stack/min_stack.py
# DIFFICULTY: MEDIUM
# ------------------
#
# Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
#
# Implement the MinStack class:
#
# MinStack() initializes the stack object.
# void push(int val) pushes the element val onto the stack.
# void pop() removes the element on the top of the stack.
# int top() gets the top element of the stack.
# int getMin() retrieves the minimum element in the stack.
#
# You must implement a solution with O(1) time complexity for each function.
#
# See https://leetcode.com/problems/min-stack
class MinStack:
    def __init__(self):
        """
        SOLUTION
        --------

        Use one stack to store the values and another stack to store the minimum values.

        It appears that the problem will not call your solution with an empty stack, so we don't need to worry about that
        case.

        COMPLEXITY
        ----------

        Time complexity is O(1) for all operations.

        Space complexity is O(n) where n is the number of elements in the stack.
        """
        self.stack: list[int] = []
        self.min_values: list[int] = []

    def push(self, value: int) -> None:
        self.stack.append(value)

        if not self.min_values:
            self.min_values.append(value)
            return

        min_value = min(value, self.getMin())
        self.min_values.append(min_value)

    def pop(self) -> None:
        self.stack.pop()
        self.min_values.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_values[-1]
